portfolio/src/main/java/com/google/sps/data/CommentAnalysis.java

```diff
@@ -0,0 +1,136 @@
+ * will eventually also analyze the sentiment of the comment text
+ */
+public class CommentAnalysis {
+  private List<ImageLabel> imageLabels;
+
+  /** Represents a single label for an image. Serves as a convenient wrapper for serialization */
+  private static class ImageLabel {
+    private final String description;
+    private final float score;
+
+    public ImageLabel(EntityAnnotation annotation) {
+      description = annotation.getDescription();
+      score = annotation.getScore();
+    }
+
+    public String getDescription() {
+      return description;
+    }
+
+    public float getScore() {
+      return score;
+    }
+  }
+
+  public CommentAnalysis() {
+    imageLabels = new ArrayList<>();
+  }
+
+  /**
+   * Attaches image analysis, reading from the given comment. Performs no analysis if image does not
+   * exist
+   */
+  public void analyzeImage(Comment comment) throws IOException {
+    if (comment.getBlobKey() == null) {
+      return;
+    }
+    byte[] imageBytes = getBlobBytes(comment.getBlobKey());
+    List<EntityAnnotation> labels = getImageLabels(imageBytes);
+    labels.stream().forEach(entity -> imageLabels.add(new ImageLabel(entity)));
+  }
+
+  @Override
+  public String toString() {
+    Gson gson = new Gson();
+    return gson.toJson(this);
+  }
+
+  /**
+   * Retrieve binary data from blobstore at the given url
+   */
+  private static byte[] getBlobBytes(BlobKey blobKey) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
+      // End index is inclusive, so we have to subtract 1 to get fetchSize bytes
+      byte[] b =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
+      outputBytes.write(b);
+
+      // If we read fewer bytes than we requested, then we reached the end
+      if (b.length < fetchSize) {
+        continueReading = false;
+      }
+
+      currentByteIndex += fetchSize;
+    }
+
+    return outputBytes.toByteArray();
+  }
+
+  /**
+   * Uses the Google Cloud Vision API to generate a list of labels that apply to the given image
+   */
+  private static List<EntityAnnotation> getImageLabels(byte[] imageBytes) throws IOException {
+    ByteString byteString = ByteString.copyFrom(imageBytes);
+    Image image = Image.newBuilder().setContent(byteString).build();
+
+    Feature feature = Feature.newBuilder().setType(Feature.Type.LABEL_DETECTION).build();
+    AnnotateImageRequest request =
+        AnnotateImageRequest.newBuilder().addFeatures(feature).setImage(image).build();
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(request);
+
+    ImageAnnotatorClient client = ImageAnnotatorClient.create();
```

Comment:
Service API calls are usually used with try-with-resources for automatic close and easy error handling.